### ISSUE TITLE
dimension error with LSTM + TimeDistributed

### ISSUE THREAD

**Issue by skeydan** [fchollet/keras#6363]
Hi,

I have the same problem as in 

https://github.com/fchollet/keras/issues/5214

and I commented there, but I'm not sure whether comments on closed issues are still getting noticed :-)

I'm really stuck and it would be great if anyone could look at that issue.

Thank you very much in advance!

Sigrid

**Comment by bstriner** [comment]
That error might refer to problems with the shape of your data, not problems with your model. Can you confirm that the target you are passing to the model is 3d? Make sure to post your xtrain and ytrain shapes and make sure they match the model summary exactly.

**Comment by skeydan** [comment]
Hi, thanks for the quick reply!
The data is just toy data - I'll paste the code below.
It's from a complete "toy notebook"
- https://github.com/skeydan/timeseries_sandbox_python/blob/master/ts_experiment.ipynb 

where I've been trying out e.g. stateful vs. stateless, and the other if branches (using LSTM + simple Dense, or stacking LSTMs) all work fine. 
The only problem is when using TimeDistributed.

Here's the code for this branch:
<pre>
lstm_num_predictions = 5
lstm_num_timesteps = 5
batch_size = 1
ts_train = np.arange(1,101, dtype='float64').reshape(-1,1) 

def create_dataset(dataset, window_size):
    dataX, dataY = [], []
    for i in range(len(dataset) - window_size):
        a = dataset[i:(i + window_size), 0]
        dataX.append(a)
        dataY.append(dataset[i + window_size, 0])
    return np.array(dataX), np.array(dataY)

X_train, y_train = create_dataset(ts_train, lstm_num_timesteps)

# reshape input to be [samples, time steps, features]
X_train = np.reshape(X_train, (X_train.shape[0], lstm_num_timesteps, lstm_num_features))

X_train.shape # (95, 5, 1)

model.add(LSTM(num_neurons,
                      #  input_dim = lstm_num_features,
                       batch_input_shape=(batch_size, X_train.shape[1], X_train.shape[2]), 
                       return_sequences = True))
model.add(TimeDistributed(Dense(lstm_num_predictions)))
model.compile(loss='mean_squared_error', optimizer='adam')

model.fit(X_train, y_train, nb_epoch = num_epochs, batch_size = batch_size)

</pre>

thx again,
Sigrid

**Comment by skeydan** [comment]
Hi again,

I think I made several mistakes at once.
The current code works, but it would be great if you could quickly check if this is the correct way to chain LSTM and Time Distributed. Many many thanks!

Here's step-by-step what I'm doing now:

1) create the original timeseries

<pre>
ts_train = np.arange(1,101, dtype='float64').reshape(-1,1)  
ts_test = np.arange(21,41, dtype='float64').reshape(-1,1)
ts_all = np.append(ts_train, ts_test).reshape(-1,1)
</pre>


2) reshape to X_train and y_train matrices with 5 timesteps each

<pre>
def create_dataset2(dataset, window_size):
    dataX, dataY = [], []
    for i in range(len(dataset) - 2 * window_size):
        a = dataset[i:(i + window_size), 0]
        #print(a)
        dataX.append(a)
        b = dataset[(i + window_size):(i + 2* window_size), 0]
        #print(b)
        dataY.append(b)
    return np.array(dataX), np.array(dataY)
    
X_train, y_train = create_dataset2(ts_train, lstm_num_timesteps)
X_test, y_test = create_dataset2(ts_test, lstm_num_timesteps)
</pre>

3) test the shapes

<pre>
X_train.shape, y_train.shape, X_test.shape,  y_test.shape
# ((90, 5), (90, 5), (10, 5), (10, 5))
</pre>


4) reshape again, to match LSTM expected input

Before, I only reshaped X, not y

<pre>
X_train = np.reshape(X_train, (X_train.shape[0], lstm_num_timesteps, lstm_num_features))
X_test = np.reshape(X_test, (X_test.shape[0], lstm_num_timesteps, lstm_num_features))
    
if lstm_predict_sequences:
    y_train = np.reshape(y_train, (y_train.shape[0], lstm_num_predictions, lstm_num_features))
    y_test = np.reshape(y_test, (y_test.shape[0], lstm_num_predictions, lstm_num_features))
        
</pre>        


5) test the shapes again

<pre>
X_train.shape, y_train.shape, X_test.shape,  y_test.shape
# (90, 5, 1), (90, 5, 1), (10, 5, 1), (10, 5, 1))
</pre>


6) create the model

<pre>
model.add(LSTM(num_neurons, batch_input_shape=(batch_size, X_train.shape[1], X_train.shape[2]), 
                       return_sequences = True))

                      
model.add(TimeDistributed(Dense(1)))
model.compile(loss='mean_squared_error', optimizer='adam')

model.fit(X_train, y_train, nb_epoch = num_epochs, batch_size = batch_size)
</pre>

7) model summary

<pre>
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
lstm_1 (LSTM)                (1, 5, 4)                 96        
_________________________________________________________________
time_distributed_1 (TimeDist (1, 5, 1)                 5         
=================================================================
Total params: 101.0
Trainable params: 101
Non-trainable params: 0.0
</pre>

thx a lot,
Sigrid

**Comment by skeydan** [comment]
>> It is weird to always have that extra 1 dimension in both the input and the target but it makes everything easier internally.

Thanks for explaining!

And thanks for the advice regarding np.expand_dims, I've always found reshape -1 somehow strange ;-)

>> In general, Error when checking model target means your numpy target arrays are not the same shapes as the output tensors.

Yes, my main problem was I somehow never realized the message was talking about the _target_ - after I'd realized that I started to understand ...;-)

Thanks a lot for your help,
Sigrid
